pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -254,6 +258,24 @@ func NewKubeGenericRuntimeManager(
 		return nil, ErrVersionNotSupported
 	}
 
+	// Check if runtime supports CDI
+	if utilfeature.DefaultFeatureGate.Enabled(features.DevicePluginCDIDevices) || utilfeature.DefaultFeatureGate.Enabled(features.DynamicResourceAllocation) {
+		minVersions := map[string]string{"containerd": minContainerdCDIVersion, "cri-o": minCRIOCDIVersion}
+		if rversion, ok := minVersions[typedVersion.RuntimeName]; ok {
+			runtimeVersion, err := kubeRuntimeManager.Version(ctx)
+			if err != nil {
+				klog.InfoS("WARNING: can't get runtime version", "error", err)
+			} else {
+				cmp, err := runtimeVersion.Compare(rversion)
+				if err != nil {
+					klog.InfoS("WARNING: can't compare versions", "runtime", typedVersion.RuntimeName, "version", runtimeVersion, "minimal version with CDI support", rversion)
+				} else if cmp < 0 {
+					klog.InfoS("WARNING: runtime doesn't support CDI", "runtime", typedVersion.RuntimeName, "version", runtimeVersion, "minimal version with CDI support", rversion)
```

Comment:
this is very concerning, there are a lot of hardcoded things here, are we sure we want to do this?

Reply:
CRI is a v1API that is meant to be general and we are coupling the kubelet to crio and containerd releases.

In addition, CDI is an alpha field depending on an alpha feature, we are adding it to a CRI v1 API, what happens if the field has to be removed or modified, how do you handle versions compatibility, do we add new checks like
if containerd 1.7.0 and crio 1.24 do this
if containerd 1.7.2 and crio 1.5 do the other?

Reply:
I agree with @aojea: detecting support for a feature based on some hard-coded version numbers is sub-optimal. Ideally, there should be a way to discover whether the feature is supported dynamically.

But even more concerning is that the warning doesn't consider whether the runtime *configuration* enables the feature. That is the much bigger problem in practice, IMHO.

As it stands, the warning or its absence are not sufficient and both cases could lead an admin down the wrong path ("no warning -> must look elsewhere" when the config disables the feature, "warning -> must upgrade runtime" when the feature got backported and the version numbers are wrong).

I'm leaning towards not merging this if we can't do better.

Reply:
> In addition, CDI is an alpha field depending on an alpha feature, we are adding it to a CRI v1 API, what >happens if the field has to be removed or modified, how do you handle versions compatibility, do we add new checks like
if containerd 1.7.0 and crio 1.24 do this
if containerd 1.7.2 and crio 1.5 do the other?

Nope, the plan was to enable a quick check now and implement proper feature discovery (Add CDI support to the RuntimeConfig CRI API) in the runtimes and replace this check with a feature check when at least CRI-O and Containerd support it.

At this stage backports are unlikely to happen. For CRI-O it doesn't make much sense as CDI is supported there for a couple of years or so. For Containerd - we tried to backport it to 1.6 and it was rejected by the project.

> As it stands, the warning or its absence are not sufficient and both cases could lead an admin down the wrong path ("no warning -> must look elsewhere" when the config disables the feature, "warning -> must upgrade runtime" when the feature got backported and the version numbers are wrong).

We can do better, but this is much longer way (see above). The idea of this PR is to do a simple check without modifying CRI API. In my opinion it's better to have at least something now and replace it with a proper feature discovery implementation than have nothing at all.

Reply:
> Nope, the plan was to enable a quick check now and implement proper feature discovery (Add CDI support to the RuntimeConfig CRI API) in the runtimes and replace this check with a feature check when at least CRI-O and Containerd support it.

I can see there are two feature gates here, is this plan discussed in any of those KEPs? I'm worried about the explosing of featuregates and dependencies created in DRA, I count at least 4 keps related with its own feature gates, so please be careful to not create circular dependencies 

> In my opinion it's better to have at least something now and replace it with a proper feature discovery implementation than have nothing at all.

we all know temporal hacks are permanent :) , I rather prefer you clearly discuss and describe this plan within the KEP so all the projects and stakeholders are aware, bear in mind you are going to create a dependency. between kubelet and containerd and crio that all the distros need to respect, this need to be properly announced at least

Reply:
What we can do now is log some information for the admin:
"DRA: container runtime is ... with version ... which may or may not support CDI. If containers get created without additional resources, double-check whether the container runtime supports CDI and whether that support is enabled."

Reply:
@aojea 
> I can see there are two feature gates here, is this plan discussed in any of those KEPs?

Which plan? To check if the runtime supports CDI? That's a minor harmless check.

> bear in mind you are going to create a dependency. between kubelet and containerd and crio

This dependency already exists. I'm not going to create anything new. Dependency to CDI is clearly stated at least in the DRA KEP.

Thank you for the review! Although I don't agree with the pushback, I'm not going to continue this discussion. Closing the issue.